torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -1415,6 +1454,130 @@ Val* distributeMul(Val* value, const VarInfo& var_info) {
   return value;
 }
 
+// This pass optimizes register usage of predicate evaluation, especially for
+// unrolled for loop.
+// For example, if I have an unrolled for loop like below:
+//   base = ...
+//   for i in 4
+//     for j in 4
+//       if base + i* stride0 + j*stride1 < T.size:
+//         expr
+// Then the above for loop uses 4 * 4 = 16 general purposed registers to
+// evaluate the lhs of the predicate.
+// However, if I transform the predicate as:
+//   base = ...
+//   lhs = base - T.size
+//   for i in 4
+//     for j in 4
+//       if lhs < -i* stride0 - j*stride1:
+//         expr
+// Then it only takes 1 general purposed register for lhs, and the rhs is a
+// compile time constant for nvRTC, so it can use the immediate fields of the
+// instruction. This optimization results in a great save on registers.
+// See also: https://github.com/csarofeen/pytorch/pull/1976
+Val* reducePredicateRegisterUsage(Val* value, const VarInfo& var_info) {
+  if (value->isABool()) {
```

Comment:
Nit: lift as `if(!value->isABool()){return value;}`

Same for the other two nested conditionals. Thanks.

Reply:
fixed